server/asset/dcr/cache.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2019, The Decred developers
+// See LICENSE for details.
+
+package dcr
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	chainjson "github.com/decred/dcrd/rpc/jsonrpc/types"
+)
+
+// The dcrBlock structure should hold a minimal amount of information about a
+// block needed to verify UTXO validity. The stake related transaction hashes
+// are stored for quick tx tree checks. More sophisticated methods such as
+// examining the pubkey scripts may be used in the future to make this structure
+// even smaller.
+type dcrBlock struct {
+	hash     chainhash.Hash
+	height   int64
+	orphaned bool
+	vote     bool // stakeholder vote result for the previous block
+	stxIDs   []chainhash.Hash
```

Comment:
For not, I'm storing the stake tree transaction hashes so that utxos classified. This adds ~ 10 txids / block to this  cache structure, so if we can find a better way to show a transaction is from the stake tree without the overhead of an the additional RPC call, I'm all for it. Peeking at pubkey scripts is probably valid, but would require some care to implement, I think.

Reply:
When you initially get the transaction data with `GetRawTransactionVerbose`, you'll have the tx hex bytes, thus you can make a `wire.MsgTx` without any more RPCs, then you can do `stake.DetermineTxType(msgTx)`.  Something like in [txhelpers]https://github.com/decred/dcrdata/blob/71410f8b5f629e844f0d740f0a658882b8221c24/txhelpers/txhelpers.go#L1058-L1079) following a `MsgTxFromHex`.